### DISCUSSION TITLE
V2 Programming Model for Azure Functions using Python (Preview)

### DISCUSSION THREAD

**Discussion by shreyabatra4** [Azure/azure-functions-python-worker#1138] [General]
The v2 programming model is designed to provide a development experience that is more familiar to Python developers. Key features include triggers and bindings declared as decorators, importing through blueprints, and support through easy to reference documentation. Note that leveraging the v2 programming model will provide an improved and seamless way to create functions, with the underlying deployment, debugging, and monitoring experience remaining the same.

To get started and create your first function, see the following articles

[Get started with VS Code](https://learn.microsoft.com/en-us/azure/azure-functions/create-first-function-vs-code-python?pivots=python-mode-decorators)
[Get started with command prompt](https://learn.microsoft.com/en-us/azure/azure-functions/create-first-function-cli-python?pivots=python-mode-decorators&tabs=azure-cli%2Cbash)
[Code examples](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-triggers-python)

To learn more about the model, see the [developer reference guide](https://learn.microsoft.com/en-us/azure/azure-functions/functions-reference-python?tabs=asgi%2Capplication-level&pivots=python-mode-decorators). Known limitations of the v2 model and recommended workarounds can be found in the [troubleshooting guide](https://learn.microsoft.com/en-us/azure/azure-functions/recover-python-functions?source=recommendations&tabs=vscode%2Cbash&pivots=python-mode-decorators).

We'd love to know your thoughts and any feedback, please let us know if you tried out the V2 model below!

**Comment by kummerer94** [comment]
I really like how this feature has taken inspiration from flask and its way of handling blueprints / routes! It will improve the DX so much and also finally make it possible to decide which functions to expose during start-up. Also, really nice how many of the triggers and output bindings are already supported.

Is there already an estimate when this will be GA?

**Reply by shreyabatra4** [reply]
Hi @kummerer94, thanks a lot for your feedback on the v2 model! 

We are tracking GA of the model for February 2023.

**Comment by MattiasAng** [comment]
Is there any documentation on Durable Functions yet?

I see supported was added but cannot get it to fully work. The client input works as expected but not the orchestration trigger.

```python
import azure.functions as func
import azure.durable_functions as df

app = df.DFApp(http_auth_level=func.AuthLevel.ANONYMOUS)


@app.route(
    route="orchestrators/{functionName}", binding_arg_name="message"
)  # Register HTTP Trigger
@app.durable_client_input(client_name="starter")
async def durable_trigger(
    req: func.HttpRequest, starter: df.DurableOrchestrationClient, message
):
    instance_id = await starter.start_new("test")
    response = starter.create_check_status_response(req, instance_id)
    message.set(response)


@app.orchestration_trigger(context_name="context", orchestration="test")
def orchestration_trigger_test(context):
    print(context)

```

Fails with the following exception `Exception: TypeError: HistoryEvent.__init__() missing 4 required positional arguments: 'EventType', 'EventId', 'IsPlayed', and 'Timestamp'`

**Reply by vrdmr** [reply]
The support isn't completely flushed out yet and we are still in the process of getting the DF design finalized, with the DF team. We'll have more updates soon.

@davidmrdavid FYI

**Reply by davidmrdavid** [reply]
Thanks for the ping @vrdmr.
@MattiasAng: preview support for DF in the new programming model hasn't been **officially** announced yet. I suspect you see it's supported from our repo's activity, but that release is one step away from completing at this time - we need to address a minor technical blocker.

**Reply by shreyabatra4** [reply]
Durable Functions is now supported for the v2 programming model. For more information and examples, see [this link](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-triggers-python#durable-functions).

**Comment by robschmuecker** [comment]
@shreyabatra4 
Are you able to indicate whether there is an implemented way of specifying retry functionality in V2 yet as there was in the `function.json` in v1?  Specifically for a timer trigger?
Ala [https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-error-pages?tabs=fixed-delay%2Cin-process&pivots=programming-language-python](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-error-pages?tabs=fixed-delay%2Cin-process&pivots=programming-language-python)
If so, could you post some code or point me to the docs please?
Thanks in advance!

**Reply by ppmdo** [reply]
Are there any news regarding this? I was planning to migrate to V2. Are retries supported?

**Reply by gavin-aguiar** [reply]
@robschmuecker @ppmdo Retries are now supported for V2 and waiting for release. Its should be available in runtime version 4.25. Tracking PR: https://github.com/Azure/azure-functions-python-worker/issues/1253

**Comment by mat-undo** [comment]
This is great, but I am not sure that it solves a key issue that I had previously: I need to be able to share proprietary python code (not available in a public repo) between multiple Azure Function apps.   As described in e.g. https://github.com/Azure/azure-functions-core-tools/issues/1987.  Does the v2 programming model solve this issue in some way that I'm not seeing?   Thanks!

**Comment by olsgaard** [comment]
Is there a recommendation from Microsoft on V1 or V2 for new projects? Are there any public depreciation plans for V1

**Reply by shreyabatra4** [reply]
Hi, the recommendation is to use v2 for new projects! We do plan to deprecate the v1 programming model in the future, but the timeline is not yet established.

**Comment by mfathi91** [comment]
@shreyabatra4 Is there anything from Python V2 model still in preview phase? In the [doc](https://learn.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-bindings?tabs=python-v2%2Cin-process%2C2x-durable-functions&pivots=programming-language-python#python-v2-programming-model) I see that durable functions with blueprints are still not fully released (seems to be in preview phase), or is the documentation out-dated?

**Reply by shreyabatra4** [reply]
Hi @mfathi91 thank you for reaching out, Durable Functions for Python v2 is indeed GA and no longer preview. I will update the documentation to reflect this.

**Comment by nadaa** [comment]
Hi,
Is there a way to create a custom decorator for an azure function?
Thanks

**Reply by gavin-aguiar** [reply]
Hi @nadaa,

You can use generic decorators if you need any decorator thats not natively supported by V2 programming model and has an extension available. Hope this helps

```

@app.function_name(name="default_template")
@app.generic_trigger(arg_name="req",
                     type="httpTrigger",
                     route="default_template")
@app.generic_output_binding(arg_name="$return", type="http")
def default_template(req: func.HttpRequest) -> func.HttpResponse:
    logging.info('Python HTTP trigger function processed a request.')

    name = req.params.get('name')
    if not name:
        try:
            req_body = req.get_json()
        except ValueError:
            pass
        else:
            name = req_body.get('name')

    if name:
        return func.HttpResponse(
            f"Hello, {name}. This HTTP triggered function "
            f"executed successfully.")
    else:
        return func.HttpResponse(
            "This HTTP triggered function executed successfully. "
            "Pass a name in the query string or in the request body for a"
            " personalized response.",
            status_code=200
        )
```

**Reply by adriennn** [reply]
@gavin-aguiar is there a way to define a fully custom decorator? e.g. a middleware function that would run between inputs and outputs? 

e.g.

```py
import azure.functions as func
from my.code import my_decorator

app = func.FunctionApp()

app.register_custom_decorator(my_decorator) # à la register_function for blueprints

@app.http_type("POST")
@app.function_name(name="my_template")
@app.route(route='my/route', trigger_arg_name='req'
@app.my_decorator() #this will just pass the req along
@app.blob_output(...)
def default_template(req: func.HttpRequest) -> func.HttpResponse:
   # Now i don't have to do any imports of custom code in this file because I do everything in my.code
```

**Reply by gavin-aguiar** [reply]
@adriennn. Currently we don't support custom decorators.

**Comment by rebremer** [comment]
Hi all, is there a timeline when V2 Programming Model will support the Kafka decorator for durable functions?

**Comment by mk7exe** [comment]
Hi. Is there a plan/timeline to add Entity Functions to V2 programming model?

**Comment by twsh15** [comment]
It'd be great if someone could help give me a hint on this ask https://github.com/Azure/azure-functions-python-worker/issues/1395

**Comment by BDGLundeSony** [comment]
is this v2 programming model ever going to get support for SignalR output bindings? It's fine and well to say we can use the generic output bindings, but there's not really any good documentation out there on how to use it to recreate the bindings that were present in the v1 model and missing in v2...

**Comment by danielnyari** [comment]
Also looking for a way to use SignalR with programming model v2

**Comment by LouisDeconinck** [comment]
Is it possible to create a binding between Azure Function Python V2 and Azure SQL Database? If so, how?
